### PR TITLE
Add WithContextAndTeardown and fix SetTeardown

### PR DESCRIPTION
**Pull request by rht**
I have removed the occurence of all SetTeardown in https://github.com/ipfs/go-ipfs/pull/1385. So this PR is basically a lib sugar.

### CODE REVIEW

File: impl-mutex.go [hunk 1]
```diff
@@ -119,6 +119,12 @@ func (p *process) SetTeardown(tf TeardownFunc) {
 	if tf == nil {
 		tf = nilTeardownFunc
 	}
+
+	select {
+	case <-p.Closed():
+		p.closeErr = tf()
```
Comment (jbenet) [inline impl-mutex.go]: maybe this should only happen once, if it was nil:

```go
if tf == nil {
  tf = nilTeardownFunc
}

p.Lock()
select {
case <-p.Closed():
  p.teardown = tf
  p.closeErr = tf()
default:
}
p.Unlock()
```

**Comment by jbenet** [comment]
LGTM, thanks

**Comment by rht** [comment]
So I will just use SetTeardown in the ipfs/go-ipfs#1385 PR?

**Comment by jbenet** [comment]
@rht probably safer to use this function... i haven't verified that this doesn't have odd edge cases. (could try testing, but this stuff is hard to test).
